src/Spanner/Database.php

```diff
@@ -0,0 +1,1313 @@
+     *
+     * Example:
+     * ```
+     * $transaction = $database->runTransaction(function (Transaction $t) use ($username, $password) {
+     *     $rows = $t->execute('SELECT * FROM Users WHERE Name = @name and PasswordHash = @password', [
+     *         'parameters' => [
+     *             'name' => $username,
+     *             'password' => password_hash($password, PASSWORD_DEFAULT)
+     *         ]
+     *     ])->rows();
+     *     $user = $rows->current();
+     *
+     *     if ($user) {
+     *         // Do something here to grant the user access.
+     *         // Maybe set a cookie?
+     *
+     *         $user['loginCount'] = $user['loginCount'] + 1;
+     *         $t->update('Users', $user);
+     *
+     *         $t->commit();
+     *     } else {
+     *         $t->rollback();
+     *     }
+     * });
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/spanner/reference/rpc/google.spanner.v1#google.spanner.v1.BeginTransactionRequest BeginTransactionRequest
+     * @see https://cloud.google.com/spanner/docs/transactions Transactions
+     * @codingStandardsIgnoreEnd
+     *
+     * @param callable $operation The operations to run in the transaction.
+     *        **Signature:** `function (Transaction $transaction)`.
+     * @param array $options [optional] {
+     *     Configuration Options
+     *
+     *     @type int $maxRetries The number of times to attempt to apply the
+     *           operation before failing. **Defaults to ** `3`.
+     *     @type Transaction $transaction If provided, the transaction will be
+     *           passed to the callable instead of attempting to begin a new
+     *           transaction.
+     * }
+     * @return mixed The return value of `$operation`.
+     */
+    public function runTransaction(callable $operation, array $options = [])
+    {
+        $options += [
+            'maxRetries' => self::MAX_RETRIES,
+            'transaction' => null
+        ];
+
+        // There isn't anything configurable here.
+        $options['transactionOptions'] = $this->configureTransactionOptions();
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READWRITE);
+
+        $attempt = 0;
+        $startTransactionFn = function ($session, $options) use (&$attempt) {
+            if ($attempt === 0 && $options['transaction'] instanceof Transaction) {
+                $transaction = $options['transaction'];
+            } elseif ($attempt === 0 && $options['transaction']) {
+                throw new \InvalidArgumentException('Given transaction must be an instance of Transaction.');
+            } else {
+                $transaction = $this->operation->transaction($session, $options);
+            }
+
+            $attempt++;
+            return $transaction;
+        };
+
+        $delayFn = function (\Exception $e) {
+            if (!($e instanceof AbortedException)) {
+                throw $e;
```

Comment:
No if it is an AbortedException, we should not auto rollback. I was talking about the case when it is some other exception.